File: docs/demo/build-banner.mjs
/**
 * Renders the project banner from `banner.html`.
 *
 *   node docs/demo/build-banner.mjs && python3 docs/demo/build-banner.py
 *
 * Chromium rather than an image library, because the banner is typography over
 * a gradient and a browser is the thing that lays type out properly. It renders
 * at twice the final size; `build-banner.py` scales the result down, which
 * comes out sharper than rendering at the final size directly.
 *
 * Requires Playwright; set PLAYWRIGHT_MODULE if it is not resolvable from here.
 */

import { mkdirSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const HERE = dirname(fileURLToPath(import.meta.url));
const OUT = join(HERE, 'banner-frames');

/**
 * The two shapes the banner is needed in.
 *
 * The wide one is the README's; the square-ish one is GitHub's social preview,
 * which is a different composition rather than the same one letterboxed — the
 * horizontal layout leaves a band of content floating in empty space at 1280 by
 * 640, so that variant stacks and centres instead.
 */
const VARIANTS = [
  { w: 1280, h: 400, unit: 63, pad: 54, gap: 52, figw: 690, out: 'banner.png' },
  { w: 1280, h: 640, unit: 68, pad: 76, gap: 40, figw: 680, out: 'banner-social.png', stacked: true },
];

async function loadPlaywright() {
  const candidates = [
    process.env.PLAYWRIGHT_MODULE,
    'playwright',
    '/opt/node22/lib/node_modules/playwright/index.mjs',
    '/usr/lib/node_modules/playwright/index.mjs',
  ].filter(Boolean);
  for (const candidate of candidates) {
    try {
      const module = await import(candidate);
      const resolved = module.chromium ? module : module.default;
      if (resolved?.chromium) return resolved;
    } catch {
      // Try the next candidate.
    }
  }
  throw new Error('Playwright not found. Install it, or set PLAYWRIGHT_MODULE.');
}

const { chromium } = await loadPlaywright();
mkdirSync(OUT, { recursive: true });

const browser = await chromium.launch({
  args: ['--no-proxy-server', '--force-color-profile=srgb', '--font-render-hinting=none'],
});

try {
  for (const variant of VARIANTS) {
    const page = await browser.newPage({
      viewport: { width: variant.w, height: variant.h },
      deviceScaleFactor: 2,
    });
    await page.goto(pathToFileURL(join(HERE, 'banner.html')).href);
    await page.addStyleTag({
      content:
        `.banner{--w:${variant.w}px;--h:${variant.h}px;--unit:${variant.unit}px;` +
        `--pad:${variant.pad}px;--gap:${variant.gap}px;--figw:${variant.figw}px}`,
    });
    if (variant.stacked) {
      await page.evaluate(() => document.querySelector('.banner').classList.add('stacked'));
    }
    // Web fonts are not used, but the gradient and the generated waveform both
    // need a frame to settle before the shot.
    await page.waitForTimeout(400);
    // The corners are rounded and must come out transparent rather than filled
    // with whatever the page behind them happened to be.
    await page.locator('.banner').screenshot({ path: join(OUT, variant.out), omitBackground: true });
    console.log(`${variant.out}  ${variant.w}x${variant.h} at 2x`);
    await page.close();
  }
} finally {
  await browser.close();
}

File: docs/demo/build-banner.py
#!/usr/bin/env python3
"""Turns the rendered banner frames into the committed assets.

Run build-banner.mjs first; this only scales and encodes what it produced.

    python3 docs/demo/build-banner.py

The two outputs are deliberately different formats.

The README's banner is WebP because the background is a smooth gradient, which
is the thing PNG compresses worst: the same image is 621 kB as a PNG and 64 kB
as WebP at quality 95, with no ringing visible around the wordmark even
magnified. GitHub renders WebP in a README.

The social preview is PNG because GitHub's uploader for it takes PNG, JPEG or
GIF and not WebP, at the 1280x640 it asks for.
"""

import pathlib
import sys

from PIL import Image

HERE = pathlib.Path(__file__).resolve().parent
FRAMES = HERE / "banner-frames"
OUT = HERE.parent / "images"

# (source, final size, destination, encoder arguments)
ASSETS = [
    ("banner.png", (1920, 600), "banner.webp", {"quality": 95, "method": 6}),
    ("banner-social.png", (1280, 640), "banner-social.png", {"optimize": True, "compress_level": 9}),
]


def main():
    if not FRAMES.is_dir():
        sys.exit("no rendered frames; run build-banner.mjs first")

    OUT.mkdir(parents=True, exist_ok=True)
    for source, size, destination, options in ASSETS:
        path = FRAMES / source
        if not path.is_file():
            sys.exit(f"{path} is missing; run build-banner.mjs first")

        # Rendered at twice this and scaled down here, which is sharper than
        # asking the browser for the final size.
        # RGBA, not RGB: the rounded corners are transparent and flattening
        # them onto white would draw four bright wedges.
        image = Image.open(path).convert("RGBA").resize(size, Image.LANCZOS)
        target = OUT / destination
        image.save(target, **options)
        print(f"{target.relative_to(HERE.parent.parent)}  {target.stat().st_size / 1024:.0f} kB  {size[0]}x{size[1]}")


if __name__ == "__main__":
    main()
